### ISSUE TITLE
Use withoutActuallyEscaping for helpers on Swift 3.1

### ISSUE THREAD

**Issue by zwaldowski** [bignerdranch/CoreDataStack#194]
Starting in Swift 3.1, the optimizer is free to introduce incorrect optimizations for the former implementation.

Before:

```swift
public func performAndWaitOrThrow<Return>(_ body: () throws -> Return) rethrows -> Return {
        func impl(execute work: () throws -> Return, recover: (Error) throws -> Void) rethrows -> Return {
            var result: Return!
            var error: Error?

            // performAndWait is marked @escaping as of iOS 10.0.
            // swiftlint:disable type_name
            typealias Fn = (() -> Void) -> Void
            let performAndWaitNoescape = unsafeBitCast(self.performAndWait, to: Fn.self)
            performAndWaitNoescape {
                do {
                    result = try work()
                } catch let e {
                    error = e
                }
            }

            if let error = error {
                try recover(error)
            }

            return result
        }

        return try impl(execute: body, recover: { throw $0 })
    }
```

After:

```swift
 public func performAndWaitOrThrow<Return>(_ body: () throws -> Return) rethrows -> Return {
        return try withoutActuallyEscaping(body) { (work) in
            var result: Return!
            var error: Error?

            performAndWait {
                do {
                    result = try work()
                } catch let e {
                    error = e
                }
            }

            if let error = error {
                throw error
            } else {
                return result
            }
        }
    }
```

**Comment by jeremy-w** [comment]
Suggests a quick grep for `unsafeBitCast` in particular and `unsafe` in general in light of continued Swift development, to catch any similar issues.

**Comment by rcedwards** [comment]
Yeah the only two occurrences are for this function in both the `NSManagedObjectContext` and `NSPersistentStoreCoordinator` extensions.

Looks good @zwaldowski I'll create a PR today.

**Comment by zwaldowski** [comment]
Re: @jeremy-w, AFAIK this should be emitting a warning under 3.1:

* https://github.com/apple/swift/blob/swift-3.1-branch/include/swift/AST/DiagnosticsSema.def#L2454
* https://github.com/apple/swift/blob/master/CHANGELOG.md

> The old workaround of using unsafeBitCast to cast to an @escaping type is not guaranteed to work in future versions of Swift, and will now raise a warning.

I'm not sure why it's not.

**Comment by jeremy-w** [comment]
@zwaldowski It's not getting tripped because we're not casting the closure from no-escape to maybe-escape, we're casting the thing calling the closure from calling a closure that might escape to calling a closure that does not. We went the other direction, at one remove.

The Swift compiler ought to emit a related diagnostic for our approach, too, so opening an issue with them would be good. I bet we're not the only ones who did it that way! Something like "changing the type of a closure argument in a function signature from from escaping to non-escaping is undefined behavior; use etc etc".

[Here's the logic used](https://github.com/apple/swift/blob/86620aaa7ebd32d33f4cdf61add5c63a72d3f02a/lib/Sema/MiscDiagnostics.cpp#L741-L748) to decide whether to emit the warning:

```c++
     if (auto fromFnTy = fromTy->getAs<FunctionType>()) {
        if (auto toFnTy = toTy->getAs<FunctionType>()) {
          // Casting a nonescaping function to escaping is UB.
          // `withoutActuallyEscaping` ought to be used instead.
          if (fromFnTy->isNoEscape() && !toFnTy->isNoEscape()) {
            TC.diagnose(DRE->getLoc(), diag::bitcasting_away_noescape,
                        fromTy, toTy);
          }
```

**Comment by rcedwards** [comment]
Merged in #195 and part of 2.2.1
